src/Stratis.Bitcoin/AddressIndexing/AddressIndexer.cs

```diff
@@ -0,0 +1,284 @@
+
+            // Collect all inputs excluding coinbases.
+            foreach (TxInList inputsCollection in block.Transactions.Where(x => !x.IsCoinBase).Select(x => x.Inputs))
+                inputs.AddRange(inputsCollection);
+
+            Transaction[] transactions = this.blockStore.GetTransactionsByIds(inputs.Select(x => x.PrevOut.Hash).ToArray());
+
+            for (int i = 0; i < inputs.Count; i++)
+            {
+                TxIn currentInput = inputs[i];
+
+                TxOut txOut = transactions[i].Outputs[currentInput.PrevOut.N];
+
+                Money amountSpent = txOut.Value;
+
+                AddressIndexData addrData = this.GetOrCreateAddressData(txOut.ScriptPubKey);
+
+                if (blockAdded)
+                {
+                    // Record money being spent.
+                    addrData.Changes.Add(new AddressBalanceChange()
+                    {
+                        Height = currentHeight,
+                        Satoshi = amountSpent.Satoshi,
+                        Deposited = false
+                    });
+                }
+                else
+                {
+                    // Remove changes.
+                    addrData.Changes.RemoveAll(x => x.Height == currentHeight);
+                }
+            }
+
+            // Process outputs.
+            foreach (Transaction tx in block.Transactions)
+            {
+                foreach (TxOut txOut in tx.Outputs)
+                {
+                    Money amountReceived = txOut.Value;
+
+                    AddressIndexData addrData = this.GetOrCreateAddressData(txOut.ScriptPubKey);
+
+                    if (blockAdded)
+                    {
+                        // Record money being sent.
+                        addrData.Changes.Add(new AddressBalanceChange()
+                        {
+                            Height = currentHeight,
+                            Satoshi = amountReceived.Satoshi,
+                            Deposited = true
+                        });
+                    }
+                    else
+                    {
+                        // Remove changes.
+                        addrData.Changes.RemoveAll(x => x.Height == currentHeight);
+                    }
+                }
+            }
+
+            this.data.TipHash = item.Value.ChainedHeader.HashBlock.ToString();
+        }
+
+        private AddressIndexData GetOrCreateAddressData(Script scriptPubKey)
+        {
+            byte[] scriptPubKeyBytes = scriptPubKey.ToBytes();
+
+            AddressIndexData addrData = this.data.AddressIndexDatas.SingleOrDefault(x => StructuralComparisons.StructuralEqualityComparer.Equals(scriptPubKeyBytes, x.ScriptPubKeyBytes));
+
+            if (addrData == null)
+            {
+                addrData = new AddressIndexData()
+                {
+                    ScriptPubKeyBytes = scriptPubKeyBytes,
+                    Changes = new List<AddressBalanceChange>()
```

Comment:
Is it not cheaper interms of mempry to just track a balance?

Reply:
Well actually this way is probably better as it makes it easier to track the height of every deposit

Reply:
it's cheaper but exchanges need a method that accepts minconf value so we need to know height of each change